### PR TITLE
Disallow running save-related operations concurrently

### LINKED ISSUE #25426: Error when attempting to save beatmap while export is in progress
Relevant logs:

```
2023-11-14 03:28:02 [error]: An unobserved error has occurred.
2023-11-14 03:28:02 [error]: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
2023-11-14 03:28:02 [error]: at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
2023-11-14 03:28:02 [error]: at osu.Game.Database.LegacyArchiveExporter`1.ExportToStream(TModel model, Stream outputStream, ProgressNotification notification, CancellationToken cancellationToken)
2023-11-14 03:28:02 [error]: at osu.Game.Database.LegacyExporter`1.<>c__DisplayClass12_0.<ExportToStreamAsync>b__1(TModel s)
2023-11-14 03:28:02 [error]: at osu.Game.Database.RealmLive`1.PerformRead(Action`1 perform)
2023-11-14 03:28:02 [error]: at osu.Game.Database.LegacyExporter`1.<>c__DisplayClass12_0.<ExportToStreamAsync>b__0()
2023-11-14 03:28:02 [error]: at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
2023-11-14 03:28:02 [error]: --- End of stack trace from previous location ---
2023-11-14 03:28:02 [error]: at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
2023-11-14 03:28:02 [error]: --- End of stack trace from previous location ---
2023-11-14 03:28:02 [error]: at osu.Game.Database.LegacyExporter`1.ExportAsync(Live`1 model, CancellationToken cancellationToken)
2023-11-14 03:28:02 [verbose]: Invalidating working beatmap cache for Jakarta Funk Brothers - FANTASTIC THREE (Erasome) [weeeeeeeeeee]
2023-11-14 03:28:02 [verbose]: ⚠️ Subsequent messages have been logged. Click to view log files.
2023-11-14 03:28:08 [error]: An unobserved error has occurred.
2023-11-14 03:28:08 [error]: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
2023-11-14 03:28:08 [error]: at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
2023-11-14 03:28:08 [error]: at osu.Game.Database.LegacyArchiveExporter`1.ExportToStream(TModel model, Stream outputStream, ProgressNotification notification, CancellationToken cancellationToken)
2023-11-14 03:28:08 [error]: at osu.Game.Database.LegacyExporter`1.<>c__DisplayClass12_0.<ExportToStreamAsync>b__1(TModel s)
2023-11-14 03:28:08 [error]: at osu.Game.Database.RealmLive`1.PerformRead(Action`1 perform)
2023-11-14 03:28:08 [error]: at osu.Game.Database.LegacyExporter`1.<>c__DisplayClass12_0.<ExportToStreamAsync>b__0()
2023-11-14 03:28:08 [error]: at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
2023-11-14 03:28:08 [error]: --- End of stack trace from previous location ---
2023-11-14 03:28:08 [error]: at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
2023-11-14 03:28:08 [error]: --- End of stack trace from previous location ---
2023-11-14 03:28:08 [error]: at osu.Game.Database.LegacyExporter`1.ExportAsync(Live`1 model, CancellationToken cancellationToken)
```

### Discussed in https://github.com/ppy/osu/discussions/25425

<div type='discussions-op-text'>

<sup>Originally posted by **hwabis** November 14, 2023</sup>
[runtime.log](https://github.com/ppy/osu/files/13344191/runtime.log)

https://github.com/ppy/osu/assets/62194990/42eed7e9-24df-4dfd-be63-064f2e4c4a7a

</div>

### PR DESCRIPTION
**Pull request by bdach**
Closes https://github.com/ppy/osu/issues/25426.

Different approach to prior ones, this just disables the relevant actions when something related to save/export is going on. Still ends up being convoluted because many things you wouldn't expect to touch save do touch save, so it's not just a concern between export and save specifically.

Can be tested manually with something like

```diff
diff --git a/osu.Game/Beatmaps/BeatmapManager.cs b/osu.Game/Beatmaps/BeatmapManager.cs
index 0610f7f6fb..e51fc75681 100644
--- a/osu.Game/Beatmaps/BeatmapManager.cs
+++ b/osu.Game/Beatmaps/BeatmapManager.cs
@@ -415,9 +415,17 @@ public void UndeleteAll()
         public Task<Live<BeatmapSetInfo>?> ImportAsUpdate(ProgressNotification notification, ImportTask importTask, BeatmapSetInfo original) =>
             beatmapImporter.ImportAsUpdate(notification, importTask, original);
 
-        public Task Export(BeatmapSetInfo beatmap) => beatmapExporter.ExportAsync(beatmap.ToLive(Realm));
+        public async Task Export(BeatmapSetInfo beatmap)
+        {
+            await Task.Delay(30000).ConfigureAwait(false);
+            await beatmapExporter.ExportAsync(beatmap.ToLive(Realm)).ConfigureAwait(false);
+        }
 
-        public Task ExportLegacy(BeatmapSetInfo beatmap) => legacyBeatmapExporter.ExportAsync(beatmap.ToLive(Realm));
+        public async Task ExportLegacy(BeatmapSetInfo beatmap)
+        {
+            await Task.Delay(30000).ConfigureAwait(false);
+            await legacyBeatmapExporter.ExportAsync(beatmap.ToLive(Realm)).ConfigureAwait(false);
+        }
 
         private void updateHashAndMarkDirty(BeatmapSetInfo setInfo)
         {

```

Only the export side can/should be tested like that, doing the same for the save side is kinda pointless since beatmap save is actually hard-sync right now, so there is no real threat of concurrent execution (but all call sites that call save need to be guarded anyway, to protect from async export).

### CODE REVIEW

File: osu.Game/Screens/Edit/Editor.cs [hunk 1]
```diff
@@ -1020,25 +1045,41 @@ private void updateLastSavedHash()
...
+        }
 
         private EditorMenuItem createExportMenu()
         {
             var exportItems = new List<MenuItem>
             {
-                new EditorMenuItem(EditorStrings.ExportForEditing, MenuItemType.Standard, () => exportBeatmap(false)) { Action = { Disabled = !RuntimeInfo.IsDesktop } },
-                new EditorMenuItem(EditorStrings.ExportForCompatibility, MenuItemType.Standard, () => exportBeatmap(true)) { Action = { Disabled = !RuntimeInfo.IsDesktop } },
+                new EditorMenuItem(EditorStrings.ExportForEditing, MenuItemType.Standard, () => exportBeatmap(false)),
+                new EditorMenuItem(EditorStrings.ExportForCompatibility, MenuItemType.Standard, () => exportBeatmap(true)),
```
Comment (bdach) [inline osu.Game/Screens/Edit/Editor.cs:1082]: These `IsDesktop` checks are lifted into `createFileMenuItems()`, as in the top-level menu item will not be added at all if we're on mobile and the beatmap cannot be exported. This is mostly so I can do the thing with dumping all related items into one `saveRelatedMenuItems` list without worrying about additional criteria for enable (thus I can just bind to the tracker's state and set disabled state directly).

**Comment by peppy** [comment]
This change is a bit hard to swallow in terms of code quality.

I'm wondering if a method like `AttemptSaveWithPostOperation(Action postOperation)` could help simplify this? where the `InProgress` check is always performed inside that method rather than at each usage.

**Review by peppy — COMMENTED** [review]
As commented.

**Comment by peppy** [comment]
I can live with this now.

**Review by peppy — APPROVED** [review]

**Pull request merged**